File: vision-be/vision-panettone/vision/network/zmq_subscriber_socket.cpp
#include "vision/network/zmq_subscriber_socket.h"

namespace vision {

ZmqSubscriberSocket::ZmqSubscriberSocket(int n_threads) :
    context_(n_threads),
    socket_(context_, ZMQ_SUB),
    fd_(socket_.get(zmq::sockopt::fd)) {}

void ZmqSubscriberSocket::connect(std::string_view address, std::span<std::string_view> topics) {
  socket_.connect(std::string{address});
  for (const auto& topic : topics) {
    socket_.set(zmq::sockopt::subscribe, topic);
  }
}

void ZmqSubscriberSocket::close() { socket_.close(); }

int ZmqSubscriberSocket::fd() const { return fd_; }

ZmqSubscriberSocket::receive_type ZmqSubscriberSocket::receive() {
  if (zmq::message_t zmq_topic; socket_.recv(zmq_topic, zmq::recv_flags::dontwait)) {
    if (zmq::message_t zmq_result; socket_.recv(zmq_result, zmq::recv_flags::dontwait)) {
      ZmqDatagram datagram;
      datagram.topic.assign(zmq_topic.to_string());
      datagram.message.assign(zmq_result.to_string());
      return datagram;
    }
  }

  return {};
}

} // namespace vision

File: tester/network/zmq_subscriber_socket.py
from typing import Optional
from network.pub_sub_mode import PubSubMode
import zmq


class ZmqSubscriberSocket:
    def __init__(self, topic: str, n_threads: int = 1, timeout_ms: int = -1):
        # Create a ZMQ Subscriber
        self.topic = topic
        self.context = zmq.Context(n_threads)
        self.timeout_ms = timeout_ms
        self.subscriber = self.context.socket(zmq.SUB)

    def connect(self, address):
        # Connect the socket to the given ZMQ address
        self.subscriber.connect(address)
        if self.timeout_ms >= 0:
            self.subscriber.setsockopt(zmq.RCVTIMEO, self.timeout_ms)
        self.subscriber.setsockopt(zmq.SUBSCRIBE, self.topic.encode())

    def receive(self, mode) -> Optional[bytes]:
        flags = zmq.DONTWAIT if mode == PubSubMode.DontWait else zmq.Flag(0)

        try:
            topic, message = self.subscriber.recv_multipart(flags=flags)
            assert topic.decode() == self.topic
            return message
        except zmq.Again:
            return None
